Fix wrong amount in old p ui - use the right amount calculation for payone too

dorin-ciceu opened the pull request:
@plentymarkets/team-order-payment

dorin-ciceu committed:
Fix amount for old payment ui

dorin-ciceu committed:
Add exchange rate, currency and calculate amount

dorin-ciceu committed:
Fix currency and amount for Payone for update payment too

estafilarakis commented on src/Services/PaymentCreation.php:
```diff
@@ -133,6 +134,26 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
+            $defaultCurrency = $currencyService->getDefaultCurrency();
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
```
If `$payment->currency != $defaultCurrency` is true, then we have a foreign currency. In this case the exchange rate should be already set (in the basket or during order creation, if we have an order). Is it necessary to change this value? Maybe you should first check it:
```php
if ($payment->exchangeRatio <= 0.0) {
    $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
}
```

dorin-ciceu replied:
I can't find any exchange ratio in the basket. The main problem here is that the payment is saved with the exchange ration = 1 if this code is not added. I also researched this on other payment methods and seems to be the right way.

estafilarakis commented on src/Services/PaymentCreation.php:
```diff
@@ -235,6 +256,26 @@ public function capturePayment(Payment $payment, Order $order)
+            $defaultCurrency = $currencyService->getDefaultCurrency();
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
```
same situation

estafilarakis reviewed (commented):
Why is this necessary? Does PayPal only accept payments in EUR and not in another currency?
Note that the default currency is not always EUR (for British systems it can be GBP and for Swiss systems CHF)!

raphael-adam commented on src/Services/PaymentCreation.php:
```diff
@@ -133,6 +134,26 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
+                $payment->amount = round(
+                    $currencyService->convertToDefaultCurrency(
+                        $payment->currency,
+                        $payment->amount,
+                        $payment->exchangeRatio
+                    ),
+                    2);
```
This looks wrong to me. I think the original amount in the foreign currency needs to be saved in the payment. Am I right @j-konopka ?

j-konopka replied:
Yes it is better to save the original amount. If their is a foreign currency you have to set the flag "isSystemCurrency" to "false"

dorin-ciceu replied:
@j-konopka @raphael-adam ok. We can save the original amount(the amount in foreign currency) but then the old payment UI is not working properly. Also, take a look at the other payment processors, all of them have the same calculation. This bug fix was added because before this fix, the amount was saved in the foreign currency. Note that if we want to save the foreign amount, we have to change it for all payment methods and another problem is that the isSystemCurrency set to false will not work.

raphael-adam replied:
Isn't the behaviour like that? :
The amount of the payment is always saved in the currency of the payment. If the payment currency is not the system currency, the exchange rate is saved and `isSystemCurrency` is set to false.
Maybe we can have a call sometime where @j-konopka can explain how the fields should be set.

j-konopka replied:
If the `isSystemCurrency` Flag is `true` the amount should be in the system currency. It could be that the currency sign is different to the system currency. But for this case an exchangeRatio should be available.
If the `isSystemCurrency` Flag is `false` the amount should be in the foreign currency and the currency sign should be also the foreign currency.

raphael-adam commented on src/Services/PaymentCreation.php:
```diff
@@ -133,6 +134,26 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
+                    ),
+                    2);
+            }
+        } catch (\Exception $cEx) {
```
Why is this necessary? These empty catch expressions are always a nightmare for debugging.

dorin-ciceu replied:
solved

dorin-ciceu committed:
Removed try catch

dorin-ciceu committed:
After CR changes - The right way
When payment is made in foreign currency in the checkout: when a payment is placed in a foreign currency, we save the foreign amount, foreign currency sign, exchange ratio and isSystemCurrency set to 0

raphael-adam commented on src/Services/PaymentCreation.php:
```diff
@@ -133,6 +134,19 @@ public function createPayment($mopId, ResponseAbstract $response, Basket $basket
+        //when a payment is placed in a foreign currency, we save the foreign amount,
+        // foreign currency sign, exchange ratio and isSystemCurrency set to 0
+        if ($payment->currency != $defaultCurrency) {
+            $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
```
I think this is not necessary, when `isSystemCurrency  = false`

raphael-adam commented on src/Services/PaymentCreation.php:
```diff
@@ -235,6 +249,26 @@ public function capturePayment(Payment $payment, Order $order)
+
+            if ($payment->currency != $defaultCurrency) {
+                $payment->exchangeRatio = $currencyService->getExchangeRatioByCurrency($payment->currency);
+                $payment->amount = round(
+                    $currencyService->convertToDefaultCurrency(
+                        $payment->currency,
+                        $payment->amount,
+                        $payment->exchangeRatio
+                    ),
+                    2);
```
Why is here still the currency convertion? And the code on top looks wrong to me: `$payment->amount = $orderData['amounts'][0]['grossTotal'];`. Maybe lets talk after the daily about this.

raphael-adam commented on src/Services/PaymentCreation.php:
```diff
@@ -235,6 +249,26 @@ public function capturePayment(Payment $payment, Order $order)
+                    ),
+                    2);
+            }
+        } catch (\Exception $cEx) {
+
+        }
```
Still the empty catch

dorin-ciceu committed:
After CR changes - get amount currency and total, remove empty try catch and simplify the code

dorin-ciceu committed:
IsSystemCurrency and exchange rate from order amount

dorin-ciceu committed:
Changelog and version added

dorin-ciceu committed:
Merge branch 'plentysystems_support' into fix/payone/wrong_amount_in_old_p_ui
